pkg/controller/seed-controller-manager/etcdbackup/etcd_backup_controller.go

```diff
@@ -0,0 +1,348 @@
+	BackendOperations
+	clock    clock.Clock
+	recorder record.EventRecorder
+}
+
+// Add creates a new Backup controller that is responsible for
+// managing cluster etcd backups
+func Add(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	snapshotDir string,
+	s3Endpoint string,
+	s3BucketName string,
+	s3AccessKeyID string,
+	s3SecretAccessKey string,
+) error {
+	log = log.Named(ControllerName)
+	client := mgr.GetClient()
+
+	reconciler := &Reconciler{
+		log:        log,
+		Client:     client,
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+		BackendOperations: &s3BackendOperations{
+			snapshotDir:       snapshotDir,
+			s3Endpoint:        s3Endpoint,
+			s3BucketName:      s3BucketName,
+			s3AccessKeyID:     s3AccessKeyID,
+			s3SecretAccessKey: s3SecretAccessKey,
+		},
+		clock: &clock.RealClock{},
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdBackup{}}, &handler.EnqueueRequestForObject{})
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	backup := &kubermaticv1.EtcdBackup{}
+	if err := r.Get(ctx, request.NamespacedName, backup); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: backup.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "backup", backup.Name)
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		kubermaticv1.ClusterConditionBackupControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(ctx, log, backup, cluster)
+		},
+	)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(backup, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ReconcilingError",
+			"failed to reconcile etcd backup %q: %v", backup.Name, err)
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, backup *kubermaticv1.EtcdBackup, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	if backup.DeletionTimestamp != nil {
+		log.Debug("Cleaning up all backups")
+		return nil, wrapErrorMessage("error cleaning up all backups: %v", r.deleteAllBackups(ctx, log, backup))
```

Comment:
I am not sure if we should delete _all_ backups when the config is removed. Maybe have a flag in the CRD spec for this?